File: app/src/main/java/androidx/preference/ClearCertPreference.java
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.

package androidx.preference;

import android.content.Context;
import android.content.DialogInterface;
import android.util.AttributeSet;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import com.ubergeek42.WeechatAndroid.R;
import com.ubergeek42.WeechatAndroid.service.SSLHandler;

import java.text.MessageFormat;

public class ClearCertPreference extends DialogPreference implements DialogFragmentGetter {

    ////////////////////////////////////////////////////////////////////////////////////////////////

    public ClearCertPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        update();
    }

    private void update() {
        final int count = SSLHandler.getInstance(getContext()).getUserCertificateCount();
        setEnabled(count > 0);
        setSummary(MessageFormat.format(getContext().getString(R.string.pref_clear_certs_summary), count));
    }

    @Override @NonNull public DialogFragment getDialogFragment() {
        return new ClearCertPreferenceFragment();
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////

    public static class ClearCertPreferenceFragment extends PreferenceDialogFragmentCompat implements DialogInterface.OnClickListener {
        protected void onPrepareDialogBuilder(AlertDialog.Builder builder) {
            super.onPrepareDialogBuilder(builder);

            builder.setTitle(null).setMessage(getString(R.string.pref_clear_certs_confirmation)).
                    setPositiveButton(getString(R.string.pref_clear_certs_positive), (dialog, which) -> {
                        boolean removed = SSLHandler.getInstance(requireContext()).removeKeystore();
                        String msg = getString(removed ?
                                R.string.pref_clear_certs_success :
                                R.string.pref_clear_certs_failure);
                        Toast.makeText(getContext(), msg, Toast.LENGTH_LONG).show();
                    }).setNegativeButton(getString(R.string.pref_clear_certs_negative), null);
        }

        @Override
        public void onDialogClosed(boolean b) {
            ((ClearCertPreference) getPreference()).update();
        }
    }
}

File: app/src/main/java/androidx/preference/ClearCertPreference.kt
package androidx.preference

import android.content.Context
import android.util.AttributeSet
import com.ubergeek42.WeechatAndroid.R
import com.ubergeek42.WeechatAndroid.service.SSLHandler
import com.ubergeek42.WeechatAndroid.utils.Toaster.Companion.ErrorToast
import com.ubergeek42.WeechatAndroid.utils.Toaster.Companion.SuccessToast

class ClearCertPreference(context: Context, attrs: AttributeSet) : ClearPreference(context, attrs) {
    override val message = R.string.pref__ClearCertPreference__prompt
    override val negativeButton = R.string.pref__ClearCertPreference__button_cancel
    override val positiveButton = R.string.pref__ClearCertPreference__button_clear

    override fun update() {
        val count = SSLHandler.getInstance(context).getUserCertificateCount()
        isEnabled = count > 0
        summary = when (count) {
            0 -> context.getString(R.string.pref__ClearCertPreference__0_entries)
            1 -> context.getString(R.string.pref__ClearCertPreference__1_entries)
            else -> context.resources.getQuantityString(R.plurals.pref__ClearCertPreference__n_entries, count, count)
        }
    }

    override fun clear() {
        val removed = SSLHandler.getInstance(context).removeUserKeystore()
        if (removed) {
            SuccessToast.show(R.string.pref__ClearCertPreference__success_cleared)
        } else {
            ErrorToast.show(R.string.pref__ClearCertPreference__error_could_not_clear)
        }
    }
}